pkg/gatewayserver/io/basicstationlns/basicstationlns.go

```diff
@@ -0,0 +1,363 @@
+					logger.WithError(err).Warn("Failed to unmarshal version message")
+					return err
+				}
+				logger = logger.WithFields(log.Fields(
+					"station", version.Station,
+					"firmware", version.Firmware,
+					"model", version.Model,
+				))
+				cfg, err := messages.GetRouterConfig(*fp, version.IsProduction(), time.Now())
+				if err != nil {
+					logger.WithError(err).Warn("Failed to generate router configuration")
+					return err
+				}
+				data, err = cfg.MarshalJSON()
+				if err != nil {
+					logger.WithError(err).Warn("Failed to marshal response message")
+					return err
+				}
+				if err := ws.WriteMessage(websocket.TextMessage, data); err != nil {
+					logger.WithError(err).Warn("Failed to send router configuration")
+					return err
+				}
+
+			case messages.TypeUpstreamJoinRequest:
+				var jreq messages.JoinRequest
+				if err := json.Unmarshal(data, &jreq); err != nil {
+					logger.WithError(err).Warn("Failed to unmarshal join-request message")
+					return nil
+				}
+				up, err := jreq.ToUplinkMessage(ids, fp.BandID, receivedAt)
+				if err != nil {
+					logger.WithError(err).Warn("Failed to parse join-request message")
+					return nil
+				}
+				if err := conn.HandleUp(up); err != nil {
+					logger.WithError(err).Warn("Failed to handle uplink message")
+				}
+				recordRTT(conn, receivedAt, jreq.RefTime)
+				atomic.StoreInt64(&latestUpstreamXTime, jreq.UpInfo.XTime)
+
+			case messages.TypeUpstreamUplinkDataFrame:
+				var updf messages.UplinkDataFrame
+				if err := json.Unmarshal(data, &updf); err != nil {
+					logger.WithError(err).Warn("Failed to unmarshal uplink data frame")
+					return nil
+				}
+				up, err := updf.ToUplinkMessage(ids, fp.BandID, receivedAt)
+				if err != nil {
+					logger.WithError(err).Warn("Failed to parse uplink data frame")
+					return nil
+				}
+				if err := conn.HandleUp(up); err != nil {
+					logger.WithError(err).Warn("Failed to handle uplink message")
+				}
+				recordRTT(conn, receivedAt, updf.RefTime)
+				atomic.StoreInt64(&latestUpstreamXTime, updf.UpInfo.XTime)
+
+			case messages.TypeUpstreamTxConfirmation:
+				var txConf messages.TxConfirmation
+				if err := json.Unmarshal(data, &txConf); err != nil {
+					logger.WithError(err).Warn("Failed to unmarshal Tx acknowledgement frame")
+					return nil
+				}
+				if cids, _, ok := s.tokens.Get(uint16(txConf.Diid), receivedAt); ok {
+					txAck := messages.ToTxAcknowledgment(cids)
+					if err := conn.HandleTxAck(&txAck); err != nil {
+						logger.WithField("diid", txConf.Diid).Warn("Failed to handle Tx acknowledgement")
+					}
+				} else {
+					logger.WithField("diid", txConf.Diid).Debug("Tx acknowledgement either does not correspond to a downlink message or arrived too late")
+				}
+				recordRTT(conn, receivedAt, txConf.RefTime)
+
+			case messages.TypeUpstreamProprietaryDataFrame:
+				return errMessageTypeNotImplemented.WithAttributes("type", typ)
```

Comment:
So we break the entire traffic handling when a gateway receives a proprietary message?